hypha/static_src/src/javascript/apply/file-uploads.js

```diff
@@ -1,57 +1,38 @@
-        // var $file_drop = $file_field.find('.form__file-drop-zone');
-
-        $file_input.on('change', function (e) {
-            e.stopPropagation();
-            e.preventDefault();
-
-            var files = $(this).prop('files');
-            var output = fileList(files);
-            $file_list.html('<ul>' + output.join('') + '</ul>');
-        });
-
-        /*
-        $file_drop.on('dragover', function (e) {
-            e.stopPropagation();
-            e.preventDefault();
-            e.dataTransfer.dropEffect = 'copy';
-        });
-
-        $file_drop.on('drop', function (e) {
-            e.stopPropagation();
-            e.preventDefault();
-            var files = e.dataTransfer.file;
-            var output = fileList(files);
-            $file_list.html('<ul>' + output.join('') + '</ul>');
-        });
-        */
-    });
-
-    function fileList(files) {
-        var output = [];
-        for (var i = 0, f, len = files.length; i < len; i++) {
-            f = files[i];
-            output.push('<li><strong>', f.name, '</strong> (', f.type || 'n/a', ') - ', formatBytes(f.size), '</li>');
+    // Initialize django-file-form
+    $('form').get().forEach(function (form) {
+        // Prevent initilising it multiple times and run it for forms
+        // that have a `form_id` field added by django-file-form.
+        if (!form.initUploadFieldsDone && form.querySelector('[name=form_id]')) {
+            init(form);
+            form.initUploadFieldsDone = true;
         }
-        return output;
-    }
+    });
 
-    function formatBytes(bytes, decimals) {
-        if (bytes === 0) {
-            return '0 Bytes';
-        }
-        var k = 1024;
-        var dm = decimals <= 0 ? 0 : decimals || 2;
-        var sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];
-        var i = Math.floor(Math.log(bytes) / Math.log(k));
-        return parseFloat((bytes / Math.pow(k, i)).toFixed(dm)) + ' ' + sizes[i];
+    function init(form) {
+        window.initUploadFields(form);
+
+        // Hide wrapper elements for hidden inputs added by django-file-form
+        $('input[type=hidden]').closest('.form__group').hide();
+
+        // For each file field in the form don't allow dropping files with an invalid file extension.
+        $('[type=file][multiple]', form).get().forEach(function (fileField) {
+            var allowedExtensions = fileField.accept.split(', ');
+
+            // TODO: this currently raises an exception when trying to remove an invalid file.
+            $(fileField).next('.dff-files').on('drop', function (e) {
+                var items = e.originalEvent.dataTransfer.items;
+                var files = e.originalEvent.dataTransfer.files;
+                for (var i = 0; i < files.length; i++) {
+                    var extension = '.' + files[i].name.split('.').slice(-1)[0];
+                    if (!allowedExtensions.includes(extension)) {
+                        items.remove(i);
```

Comment:
Only thing that is pending now is the client side validation of the files with invalid extension. This doesn't seem to work and gives error:
```
Failed to execute 'remove' on 'DataTransferItemList': The list is not writable.
```
Unfortunately, we can not remove items from drop box like this. Looked over the internet for a similar possible solution but couldn't find much. I have also opened an issue related to this at https://github.com/mbraak/django-file-form/issues/317, maybe we could show a small warning and let user remove files with invalid extension, if there is no way to remove those files automatically.

Reply:
A warning sounds like an excellent solution to this problem. Friendlier then an error on submit.

Reply:
Besides a warning simply `.preventDefault()` might work too. I guess it would just cancel all files that are being dropped.

Reply:
`. preventDefault()` also doesn't seem to work here.

Reply:
Added a warning on invalid file uploads for now. Maybe we can fix it at django-file-form later on to have those warning on specific files.